util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+    """Convert a top-level configuration and IP register definition set
+    into a System View Description ("SVD") file. SVD is an XML format,
+    originally defined by ARM to "formalize the description of the system
+    contained in Arm Cortex-M processor-based microcontrollers".¹ It has
+    since been extended for use on other platforms, and is used in other
+    fields such as Rust embedded programming.²
+
+    SVD is an XML-based format with standardized definitions for memory
+    layouts, register definitions, and similar microcontroller fields.
+
+    The input HJSON must be a top-level configuration (earlgrey) and the
+    full set of IP modeule definitions for that configuration. These two
+    will be parsed and converted into a single XML tree with a top-level
+    "<device>" node. Both the top configuration and IP modules must have
+    been previously validated. (see ??? and `reggen.validate()`)
+
+    If `validate` is true then the resulting SVD file is validated using
+    the `pysvd` module.
+
+    ¹ http://www.keil.com/pack/doc/CMSIS/SVD/html/index.html
+    ² https://github.com/rust-embedded/svd2rust"""
+
+    root = device(top, ips, version)
+
+    # Simply constructing a Device is enough to run the pysvd parser and
+    # validate the structure of the XML tree.
+    if validate:
+        pysvd.element.Device(root)
+
+    # Workaround for limitations of SVD/pysvd. See comment in `cpu()`.
+    root.find('.cpu/name').text = 'RISCV'
+
+    beautify(root)
+    return root
+
+def write_to_file(device: ET.Element, copyright: [str], output):
+    """Write out the generated SVD <device> element and its children to a
+    file in XML format. The SVD contents are preceeded by a comment with
+    the given copyright notice and a warning that the file was generated
+    by this script."""
+
+    def write_element(element, xml):
+        et = ET.ElementTree(element)
+        et.write(output, encoding='unicode', xml_declaration=xml)
+
+    comment = ET.Comment('\n'.join(('', *copyright, GENERATED_NOTICE, '')))
+    comment.tail = '\n'
+
+    # The python3 ElementTree API doesn't natively support a top-level
+    # documentation comment. Fake it for reader clarity.
+    write_element(comment, True)
+    write_element(device, False)
+
+def validate(top: hjson, ips: {"name": hjson}) -> bool:
+    """Validate the combined top-level HJSON and IP module definitions.
+    This checks for required fields in all the HJSON and ensures that a
+    corresponding IP block exists for all modules in `top`.
+
+    This assumes each individual IP module was previously validated using
+    the `reggen.validate.validate()` function."""
+
+    for field in ['name', 'datawidth', 'module']:
+        if not field in top:
+            yield 'missing required field %s in top hjson' % field
+
+    for (i, module) in enumerate(top['module']):
+        name = module.get('name') or ordinal(i)
```

Comment:
I don't see the benefit of having `ordinal` here, I'd drop the `ordinal` method and just call out the index.